File: frontend/src/lib/models/discount.js
import { ValidationModel } from "./validation";
import { Discounts } from "../database/discounts";

function DiscountModel(discount = {}) {
  const constraints = {
    name: {
      presence: true,
      type: "string",
      length: {
        minimum: 1,
        message: "Please enter a name.",
      },
    },
    active: {
      presence: true,
      type: "boolean",
    },
    percentage: {
      presence: false,
      type: "number",
    },
    amount: {
      presence: false,
      type: "number",
    },
    startAt: {
      presence: true,
    },
    endAt: {
      presence: false,
    },
    minimumPrice: {
      presence: false,
      type: "number",
    },
    minimumQuantity: {
      presence: false,
      type: "number",
    },
    userId: {
      presence: false,
      type: "string",
    },
    categoryId: {
      presence: false,
      type: "string",
    },
    productId: {
      presence: false,
      type: "string",
    },
  };

  const validation = ValidationModel(discount, constraints);
  const fields = validation.fields();

  function add() {
    return Discounts().add(validValues());
  }

  function edit(discountId) {
    return Discounts().edit({
      ...validValues(),
      id: discountId,
    });
  }

  function find(params) {
    return Discounts().find(params);
  }

  function findOne(discountId) {
    return Discounts().findOne(discountId);
  }

  function remove(discountId) {
    return Discounts().remove(discountId);
  }

  function valid() {
    return validation.valid(fields);
  }

  function validValues() {
    return validation.validValues(fields);
  }

  return Object.freeze({
    add,
    edit,
    fields,
    find,
    findOne,
    remove,
    valid,
    validValues,
  });
}

export { DiscountModel };

File: backend/app/graphql/types/discount.rb
# frozen_string_literal: true

module Types
  class Discount < Types::BaseObject
    field :id, String, null: false, method: :uuid
    field :name, String, null: false
    field :product, Types::Product, null: true
    field :category, Types::Category, null: true
    field :user, Types::User, null: true
    field :productId, String, null: true
    field :categoryId, String, null: true
    field :userId, String, null: true
    field :active, Boolean, null: true
    field :percentage, Float, null: true
    field :amount, Integer, null: true
    field :start_at, GraphQL::Types::ISO8601DateTime, null: false
    field :end_at, GraphQL::Types::ISO8601DateTime, null: true
    field :minimum_quantity, Float, null: true
    field :minimum_price, Float, null: true
  end
end
